uint64 rangeFront
uint64 rangeBack
uint64 rangeRight
uint64 rangeLeft
